controllers/auth.js

```diff
@@ -1,83 +1,142 @@
-    if ( req.body.password.length <= 7 )
-        return res.status( 400 ).send( 'Invalid password' );
+    if ( !validEmail( req.body.email ) )
+        return res.status( 400 ).json( { error: 'invalid email' } );
 
-    User.findOne( { email: req.body.email } ).exec()
-    .then( ( user ) => {
-        if ( !user ) return res.status( 404 ).send( 'No user with that email' );
+    let sendToken = user => {
+        if ( !user )
+            return res.status( 404 ).json( { error: 'no user with that email' } );
 
+        // compare passwords
         user.comparePassword( req.body.password, ( err, isMatch ) => {
             if ( err ) return next( err );
-            if ( !isMatch ) return res.status( 401 ).send( 'Incorrect password' );
 
-            let payload = {
-                id: user._id,
-                firstName: user.firstName,
-                lastName: user.lastName,
-                username: user.username,
-                email: user.email,
-                profPic: user.profPic,
-                isAdmin: !!user.isAdmin
-            };
+            if ( !isMatch )
+                return res.status( 401 ).json( { error: 'incorrect password' } );
+
+            let payload = { id: user._id };
 
-            let token = jwt.encode( payload, config.secretKey );
+            let token = jwt.sign( payload, secretKey );
 
             user.token = token;
 
-            user.save().then( () => {
-                res.json( { token: token } );
-            }).catch( ( err ) => next( err ) )
+            user.save()
+                .then( () => res.json( { token: token } ) )
+                .catch( err => next( err ) );
         });
-    }).catch( ( err ) => next( err ) );
+    }
+
+    // send a token to user
+    User.findOne( { email: req.body.email } ).exec()
+        .then( sendToken )
+        .catch( err => next( err ) );
+}
+
+exports.resetPassword = ( req, res, next ) => {
+    // TODO reset password
+}
+
+exports.forgotPassword = ( req, res, next ) => {
+    // TODO forgot password
+}
+
+exports.tokenRequired = ( req, res, next ) => {
+    validateToken( req, res, next );
 }
 
 exports.adminRequired = ( req, res, next ) => {
-    var token = req.headers[ 'x-access-token' ];
+    validateToken( req, res, next, true );
+}
 
+exports.superAdminRequired = ( req, res, next ) => {
+    validateToken( req, res, next, false, true );
+}
+
+exports.leaderRequired = ( req, res, next ) => {
+    // TODO leader required
+}
+
+const validateToken = ( req, res, next, admin, superAdmin ) => {
+    let token = req.headers[ 'x-access-token' ] || req.body.token;
+
+    if ( !token )
+        return res.status( 403 ).json( { error: 'token required' } );
+
+    let decoded;
+
+    // try verify user's token
     try {
-        var decoded = jwt.decode( token, config.secretKey, true );
-    } catch ( err ) {
-        return res.status( 403 ).send( 'Failed to authenticate token' );
+        decoded = jwt.verify( token, secretKey );

```

Comment:
Didn't you change this already ?

Reply:
What do you mean?

Reply:
JWT > bcrypt

Reply:
`jwt` for generate a token, whereas `bcrypt` for hashing passwords.
Transition:
`bcrypt` > `bcrypt-nodejs`
`jsonwebtoken` > `jwt-simple`